File: auth0_guardian_android/android/src/main/kotlin/com/kpler/auth0_guardian_flutter/GuardianFlutter.kt
package com.kpler.auth0_guardian_flutter

import EnrolledDevice
import android.net.Uri
import com.auth0.android.guardian.sdk.CurrentDevice
import com.auth0.android.guardian.sdk.Enrollment
import com.auth0.android.guardian.sdk.Guardian
import com.auth0.android.guardian.sdk.ParcelableNotification
import com.auth0.android.guardian.sdk.networking.Callback
import com.kpler.auth0_guardian_flutter.utils.SigningKeyService
import com.kpler.auth0_guardian_flutter.utils.getGuardianErrorMessage
import com.kpler.auth0_guardian_flutter.utils.toFlutterResult
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler

class GuardianFlutter : MethodCallHandler {
    private val methodChannelName = "com.kpler/auth0_guardian_android/guardian"

    private var channel: MethodChannel

    private var plugin: KotlinFlutterPlugin?

    constructor(plugin: KotlinFlutterPlugin) {
        this.plugin = plugin
        this.channel = MethodChannel(plugin.flutterPluginBinding.binaryMessenger, methodChannelName)
        this.channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        val args = call.arguments as Map<String, Any>
        when (call.method) {
            "enroll" -> enroll(args, result)
            "acceptRequest" -> acceptRequest(args, result)
            "rejectRequest" -> rejectRequest(args, result)
            "isGuardianNotification" -> isGuardianNotification(args, result)
            "generateTOTP" -> generateTOTPCode(args, result)
            else -> result.notImplemented()
        }
    }

    private fun buildGuardian(domain: String): Guardian {
        val url = Uri.parse(domain)
        return Guardian.Builder().url(url).build()
    }

    private fun enroll(args: Map<String, Any>, result: MethodChannel.Result) {
        try {
            // Extract the arguments from the method call (Flutter).
            val forDomain = args["forDomain"] as String
            val usingUri = args["usingUri"] as String
            val notificationToken = args["notificationToken"] as? String

            // Create the device and key pair objects.
            val device = CurrentDevice(plugin!!.context, notificationToken, "Android device")
            val keyPair = SigningKeyService(forDomain).generateSigningKey()

            // Define the callback to handle the enrollment result.
            val callback =
                    object : Callback<Enrollment> {
                        override fun onSuccess(enrollment: Enrollment) {
                            result.success(enrollment.toFlutterResult())
                        }
                        override fun onFailure(exception: Throwable) {
                            result.error(
                                    "EnrollFailure",
                                    getGuardianErrorMessage(
                                            "GuardianFlutter.enroll() enrollment failed"
                                    ),
                                    exception.message
                            )
                        }
                    }

            // Start the enrollment process.
            buildGuardian(forDomain).enroll(usingUri, device, keyPair).start(callback)
        } catch (exception: Exception) {
            result.error(
                    "UnexpectedError",
                    getGuardianErrorMessage("GuardianFlutter.enroll() thrown an unexpected error"),
                    exception.message
            )
        }
    }

    private fun acceptRequest(args: Map<String, Any>, result: MethodChannel.Result) {
        try {
            // Extract the arguments from the method call (Flutter).
            val payload = args["notification"] as Map<String, String>
            val forDomain = args["forDomain"] as String

            // Generate the notification object from the push payload.
            val notification: ParcelableNotification? = Guardian.parseNotification(payload)

            // Get the signing key.
            val signingKey = SigningKeyService(forDomain).getSigningKey()
            val enrollment: Enrollment = EnrolledDevice(signingKey)

            // Define the callback to handle the acceptRequest result.
            val callback =
                    object : Callback<Void> {
                        override fun onSuccess(response: Void?) {
                            result.success(true)
                        }
                        override fun onFailure(exception: Throwable) {
                            result.error(
                                    "EnrollFailure",
                                    getGuardianErrorMessage(
                                            "GuardianFlutter.acceptRequest() failed"
                                    ),
                                    exception.message
                            )
                        }
                    }

            // Start the enrollment process.
            buildGuardian(forDomain).allow(notification!!, enrollment).start(callback)
        } catch (exception: Exception) {
            result.error(
                    "UnexpectedError",
                    getGuardianErrorMessage(
                            "GuardianFlutter.acceptRequest() thrown an unexpected error"
                    ),
                    exception.message
            )
        }
    }

    private fun rejectRequest(args: Map<String, Any>, result: MethodChannel.Result) {
        try {
            // Extract the arguments from the method call (Flutter).
            val payload = args["notification"] as Map<String, String>
            val withReason = args["withReason"] as? String?
            val forDomain = args["forDomain"] as String

            // Generate the notification object from the push payload.
            val notification: ParcelableNotification? = Guardian.parseNotification(payload)

            // Get the signing key.
            val signingKey = SigningKeyService(forDomain).getSigningKey()
            val enrollment: Enrollment = EnrolledDevice(signingKey)

            // Define the callback to handle the acceptRequest result.
            val callback =
                    object : Callback<Void> {
                        override fun onSuccess(response: Void?) {
                            result.success(true)
                        }
                        override fun onFailure(exception: Throwable) {
                            result.error(
                                    "EnrollFailure",
                                    getGuardianErrorMessage(
                                            "GuardianFlutter.rejectRequest() failed"
                                    ),
                                    exception.message
                            )
                        }
                    }

            // Start the enrollment process.
            buildGuardian(forDomain).reject(notification!!, enrollment, withReason).start(callback)
        } catch (exception: Exception) {
            result.error(
                    "UnexpectedError",
                    getGuardianErrorMessage(
                            "GuardianFlutter.rejectRequest() thrown an unexpected error"
                    ),
                    exception.message
            )
        }
    }

    private fun isGuardianNotification(args: Map<String, Any>, result: MethodChannel.Result) {
        try {
            val payload = args["notification"] as Map<String, String>
            val notification: ParcelableNotification? = Guardian.parseNotification(payload)
            result.success(notification != null)
        } catch (exception: Exception) {
            result.error(
                    "UnexpectedError",
                    getGuardianErrorMessage(
                            "GuardianFlutter.isGuardianNotification() thrown an unexpected error"
                    ),
                    exception.message
            )
        }
    }

    private fun generateTOTPCode(args: Map<String, Any>, result: MethodChannel.Result) {
        try {
            val generator = LoginCodeGenerator()
            val enrollmentCode = args["enrollmentCode"] as String
            val code = generator.getCode(enrollmentCode)
            result.success(code)
        } catch (exception: Exception) {
            result.error(
                "UnexpectedError",
                getGuardianErrorMessage(
                    "GuardianFlutter.generateCode() thrown an unexpected error"
                ),
                exception.message
            )
        }
    }

    fun dispose() {
        channel.setMethodCallHandler(null)
        plugin = null
    }
}

File: auth0_guardian_ios/ios/Classes/GuardianFlutter.swift
import Flutter
import Foundation
import Guardian

public class GuardianFlutter: NSObject {
  static let METHOD_CHANNEL_NAME = "com.kpler/auth0_guardian_ios/guardian"

  var plugin: SwiftFlutterPlugin?

  init(plugin: SwiftFlutterPlugin) {
    super.init()
    self.plugin = plugin
    let channel = FlutterMethodChannel(
      name: GuardianFlutter.METHOD_CHANNEL_NAME,
      binaryMessenger: plugin.registrar!.messenger()
    )
    channel.setMethodCallHandler(handle(_:result:))
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let arguments = call.arguments as? NSDictionary
    let args = call.arguments as! [String: Any]

    switch call.method {
    case "enroll":
      enroll(args: args, result: result)
    case "acceptRequest":
      acceptRequest(args: args, result: result)
    case "rejectRequest":
      rejectRequest(args: args, result: result)
    case "isGuardianNotification":
      isGuardianNotification(args: args, result: result)
    case "generateTOTP":
      generateTOTP(args: args, result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  public func enroll(args: [String: Any], result: @escaping FlutterResult) {
    do {
      // Extract the arguments from the method call (Flutter).
      let forDomain = args["forDomain"] as! String
      let usingUri = args["usingUri"] as! String
      let notificationToken = args["notificationToken"] as! String

      // Get the signing key and the verification key.
      let signingKey = try SigningKeyService(domain: forDomain).getSigningKey()
      let verificationKey = try signingKey.verificationKey()

      // Trigger the enrollment.
      Guardian.enroll(
        forDomain: forDomain,
        usingUri: usingUri,
        notificationToken: notificationToken,
        signingKey: signingKey,
        verificationKey: verificationKey
      ).start { response in
        switch response {
        case .success(let enrollment):
          result(enrollment.toFlutterResult())
          return
        case .failure(let error):
          result(
            getGuardianFlutterError(error, "GuardianFlutter.enroll() enrollment failed")
          )
          return
        }
      }
    } catch (let error) {
      result(
        getGuardianFlutterError(error, "GuardianFlutter.enroll() thrown an unexpected error")
      )
    }
  }

  public func acceptRequest(args: [String: Any], result: @escaping FlutterResult) {
    do {
      // Extract the arguments from the method call (Flutter).
      let payload = args["notification"] as! [AnyHashable: Any]
      let forDomain = args["forDomain"] as! String

      // Generate the notification object from the push payload.
      let notification = Guardian.notification(from: payload)

      // Get the signing key and the verification key.
      let signingKey = try SigningKeyService(domain: forDomain).getSigningKey()

      // Trigger the authentication request acceptance.
      Guardian
        .authentication(forDomain: forDomain, device: signingKey.getDevice())
        .allow(notification: notification!)
        .start { response in
          switch response {
          case .success:
            result(true)
            return
          case .failure(let error):
            result(
              getGuardianFlutterError(error, "GuardianFlutter.acceptRequest() failed")
            )
            return
          }
        }
    } catch (let error) {
      result(
        getGuardianFlutterError(
          error, "GuardianFlutter.acceptRequest() throwed an unexpected error"
        )
      )
    }
  }

  public func rejectRequest(args: [String: Any], result: @escaping FlutterResult) {
    do {
      // Extract the arguments from the method call (Flutter).
      let payload = args["notification"] as! [AnyHashable: Any]
      let withReason = args["withReason"] as? String? ?? nil
      let forDomain = args["forDomain"] as! String

      // Generate the notification object from the push payload.
      let notification = Guardian.notification(from: payload)

      // Get the signing key and the verification key.
      let signingKey = try SigningKeyService(domain: forDomain).getSigningKey()

      // Trigger the authentication request acceptance.
      Guardian
        .authentication(forDomain: forDomain, device: signingKey.getDevice())
        .reject(notification: notification!, withReason: withReason)
        .start { response in
          switch response {
          case .success:
            result(true)
            return
          case .failure(let error):
            result(
              getGuardianFlutterError(error, "GuardianFlutter.rejectRequest() failed")
            )
            return
          }
        }
    } catch (let error) {
      result(
        getGuardianFlutterError(
          error, "GuardianFlutter.rejectRequest() throwed an unexpected error"
        )
      )
    }
  }

  public func isGuardianNotification(args: [String: Any], result: @escaping FlutterResult) {
    let payload = args["notification"] as! [AnyHashable: Any]
    let notification = Guardian.notification(from: payload)
    result(notification != nil)
  }

  public func generateTOTP(args: [String: Any], result: @escaping FlutterResult) {
    do {
      let enrollmentCode = args["enrollmentCode"] as! String
      let codeGenerator = try Guardian.totp(
        base32Secret: enrollmentCode,
        algorithm: .sha1
      )
      let code = codeGenerator.code()
      result(String(code))
    } catch (let error) {
      result(
        getGuardianFlutterError(
          error, "GuardianFlutter.generateTOTP() throwed an unexpected error"
        )
      )
    }
  }

  public func dispose() {
    plugin = nil
  }

  deinit {
    dispose()
  }
}
